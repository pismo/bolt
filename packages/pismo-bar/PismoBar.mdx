---
'name': PismoBar
---

import { Fragment } from 'react'
import { Playground } from 'docz'
import { PismoBar } from './PismoBar'
import { Bolt } from '@pismo/bolt-core'

# PismoBar

A **PismoBar** expõe a barra multiaplicativos da _Pismo_ contendo os ícones e url de todas as suas aplicações, é uma maneira rápida do usuário navegar por todos os aplicativos fornecidos pela _Pismo_

### Propriedades

| **prop**     | **type**                                                               | **required** | **desc**                                                                                       |
| ------------ | ---------------------------------------------------------------------- | :----------: | ---------------------------------------------------------------------------------------------- |
| AppBarProps  | AppBarProps                                                            |    false     | propriedades aplicadas à [_AppBar_](https://material-ui.com/pt/api/app-bar/)                   |
| ToolbarProps | ToolbarProps                                                           |    false     | propriedades aplicadas à [_ToolbarProps_](https://material-ui.com/pt/api/toolbar/#toolbar-api) |
| current      | `'Marketplace' / 'Adjustments' / 'Customer' / 'Backoffice' / 'WebPOS'` |     true     | variável que deve indicar a aplicação atual                                                    |
| appSelected  | ({ url: string }) => void                                              |    false     | Método executado quando uma aplicação é selecionada. Retorna a url da aplicação                |
| contract     | boolean                                                                |    false     | forma contraída da barra                                                                       |
| onClick      | () => void                                                             |    false     | se existir essa função será invocada uma vez antes de abrir a barra de apps                    |

<Playground>
  <Bolt>
    <PismoBar AppBarProps={{ position: 'relative' }} current='marketplace' />
  </Bolt>
</Playground>
