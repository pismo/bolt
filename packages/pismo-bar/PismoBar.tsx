import * as React from 'react'

import Box from '@material-ui/core/Box'
import { makeStyles, Theme } from '@material-ui/core/styles'
import Drawer from '@material-ui/core/Drawer'
import Grid from '@material-ui/core/Grid'
import useMediaQuery from '@material-ui/core/useMediaQuery'
import Fab from '@material-ui/core/Fab'

import CloseIcon from '@material-ui/icons/Close'

import {
  PismoBarProps,
  ApplicationsType,
  maxWidth
} from './components/interfaces'
import { ApplicationToolbar } from './components/ApplicationToolbar'
import { ApplicationButton } from './components/ApplicationButton'

const { Fragment, useState } = React

const applications: ApplicationsType = {
  marketplace: {
    name: 'Marketplace',
    url: ''
  },
  adjustments: {
    name: 'Adjustments',
    url: ''
  },
  customer: {
    name: 'Customer',
    url: ''
  },
  backoffice: {
    name: 'Backoffice',
    url: ''
  },
  webpos: {
    name: 'WebPOS',
    url: ''
  }
}

const useStyles = makeStyles((theme: any) => {
  const colors = theme.palette.colors
  return {
    drawer: {
      maxWidth,
      height: '100vh'
    },
    drawerPaper: {
      maxWidth,
      height: '100vh',
      overflow: 'hidden',
      backgroundColor: theme.palette.background.default,
      color: colors.text['50'],
      boxShadow: 'unset'
    },

    appButtonContainer: {
      padding: '10px'
    },

    closeButton: {
      minHeight: '0px',
      width: '36px',
      height: '36px',
      '&:hover': {
        backgroundColor: colors.background['60']
      },
      '& svg': {
        color: '#fff'
      }
    }
  }
})

const PismoBar: React.FC<PismoBarProps> = ({
  AppBarProps,
  ToolbarProps,
  current,
  appSelected,
  contract,
  full,
  onClick
}: PismoBarProps) => {
  const classes = useStyles({})
  const matches = useMediaQuery((theme: Theme) => theme.breakpoints.down('xs'))

  const [isOpen, setIsOpen] = useState<boolean>(false)
  const [countClick, setCountClick] = useState<number>(0)

  const toOpen = () => {
    if (onClick && countClick === 0) {
      setCountClick(1)
      onClick()
    } else {
      setCountClick(0)
      setIsOpen(true)
    }
  }

  const toClose = () => {
    setIsOpen(false)
  }

  const appSelectedHandler = ({ url }) => {
    if (appSelected) {
      appSelected({ url })
    }
  }

  return (
    <Fragment>
      <ApplicationToolbar
        applications={applications}
        AppBarProps={AppBarProps}
        ToolbarProps={ToolbarProps}
        current={current}
        onClick={toOpen}
        contract={contract}
        full={full}
      />
      <Drawer
        className={`Bolt-PismoBar-drawer`}
        classes={{ root: classes.drawer, paper: `${classes.drawerPaper}` }}
        anchor={matches || contract ? 'left' : 'top'}
        open={isOpen}
        onClose={toClose}
        ModalProps={{ hideBackdrop: true }}
        data-testid='drawer'
      >
        <Box textAlign='center' py='15px'>
          <Fab
            variant='round'
            color='secondary'
            className={classes.closeButton}
            onClick={toClose}
          >
            <CloseIcon />
          </Fab>
        </Box>
        <Grid
          container
          spacing={2}
          className={`${
            classes.appButtonContainer
          } Bolt-PismoBar-appButtonContainer`}
        >
          {Object.keys(applications).map((k, index) => (
            <ApplicationButton
              data={applications[k]}
              isSelected={k === current}
              key={`${k}-${index}`}
              onClick={appSelectedHandler}
            />
          ))}
        </Grid>
        <Box mt='auto'>
          <ApplicationToolbar
            applications={applications}
            AppBarProps={AppBarProps}
            ToolbarProps={ToolbarProps}
            current={current}
            onClick={toClose}
            full
          />
        </Box>
      </Drawer>
    </Fragment>
  )
}

export { PismoBar }
