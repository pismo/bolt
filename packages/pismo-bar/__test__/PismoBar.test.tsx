import {
  cleanup,
  render,
  fireEvent,
  wait,
  getByTestId
} from '@testing-library/react'
import 'jest-dom/extend-expect'
import * as React from 'react'

import { Bolt } from '@pismo/bolt-core'
import { PismoBar } from '../PismoBar'

afterEach(cleanup)

describe('@pismo/bolt-pismo-bar', () => {
  test('it should display the bar', () => {
    const Component = ({ current }) => {
      return (
        <Bolt>
          <PismoBar AppBarProps={{ position: 'relative' }} current={current} />
        </Bolt>
      )
    }
    const testOne = render(<Component current='marketplace' />)
    expect(testOne.container).toMatchSnapshot()

    expect(testOne.getByTestId('Marketplace')).toBeDefined()

    const testTwo = render(<Component current='backoffice' />)

    expect(testTwo.getByTestId('Backoffice')).toBeDefined()
  })

  test('when you click the button you should open the icon bar', async () => {
    const Component = ({ current }) => {
      return (
        <Bolt>
          <PismoBar AppBarProps={{ position: 'relative' }} current={current} />
        </Bolt>
      )
    }
    const { container } = render(<Component current='marketplace' />)

    const button = getByTestId(container, 'mainButton')
    fireEvent.click(button)

    await wait(() => {}, { timeout: 1000 })

    const drawer = getByTestId(container.parentElement as HTMLElement, 'drawer')

    expect(drawer).toBeDefined()
    expect(container.parentElement).toMatchSnapshot()
  })
})
