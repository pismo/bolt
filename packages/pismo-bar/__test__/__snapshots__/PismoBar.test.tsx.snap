// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@pismo/bolt-pismo-bar it should display the bar 1`] = `
<div>
  <header
    class="MuiPaper-root MuiPaper-elevation4 MuiAppBar-root MuiAppBar-positionRelative makeStyles-bar-4 makeStyles-bar-7 MuiAppBar-colorPrimary"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular makeStyles-toolbar-5 makeStyles-toolbar-8 MuiToolbar-gutters"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root"
        data-testid="mainButton"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            data-testid="Marketplace"
            focusable="false"
            role="presentation"
            viewBox="0 0 16 16"
          >
            <path
              d="M0 4H4V0H0V4ZM6 16H10V12H6V16ZM4 16H0V12H4V16ZM0 10H4V6H0V10ZM10 10H6V6H10V10ZM12 0V4H16V0H12ZM10 4H6V0H10V4ZM12 10H16V6H12V10ZM16 16H12V12H16V16Z"
              fill="currentColor"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </header>
</div>
`;

exports[`@pismo/bolt-pismo-bar when you click the button you should open the icon bar 1`] = `
<body
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation4 MuiAppBar-root MuiAppBar-positionRelative makeStyles-bar-97 makeStyles-bar-100 MuiAppBar-colorPrimary"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular makeStyles-toolbar-98 makeStyles-toolbar-101 MuiToolbar-gutters"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root"
          data-testid="mainButton"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              data-testid="Marketplace"
              focusable="false"
              role="presentation"
              viewBox="0 0 16 16"
            >
              <path
                d="M0 4H4V0H0V4ZM6 16H10V12H6V16ZM4 16H0V12H4V16ZM0 10H4V6H0V10ZM10 10H6V6H10V10ZM12 0V4H16V0H12ZM10 4H6V0H10V4ZM12 10H16V6H12V10ZM16 16H12V12H16V16Z"
                fill="currentColor"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </header>
  </div>
  <span
    aria-hidden="true"
    style="display: block; position: absolute; top: -9999px; left: -9999px; font-size: 300px; width: auto; height: auto; line-height: normal; margin: 0px; padding: 0px; font-variant: normal; white-space: nowrap; font-family: Lato,serif; font-style: normal; font-weight: 400;"
  >
    BESbswy
  </span>
  <span
    aria-hidden="true"
    style="display: block; position: absolute; top: -9999px; left: -9999px; font-size: 300px; width: auto; height: auto; line-height: normal; margin: 0px; padding: 0px; font-variant: normal; white-space: nowrap; font-family: Lato,sans-serif; font-style: normal; font-weight: 400;"
  >
    BESbswy
  </span>
  <span
    aria-hidden="true"
    style="display: block; position: absolute; top: -9999px; left: -9999px; font-size: 300px; width: auto; height: auto; line-height: normal; margin: 0px; padding: 0px; font-variant: normal; white-space: nowrap; font-family: serif; font-style: normal; font-weight: 400;"
  >
    BESbswy
  </span>
  <span
    aria-hidden="true"
    style="display: block; position: absolute; top: -9999px; left: -9999px; font-size: 300px; width: auto; height: auto; line-height: normal; margin: 0px; padding: 0px; font-variant: normal; white-space: nowrap; font-family: sans-serif; font-style: normal; font-weight: 400;"
  >
    BESbswy
  </span>
  <div
    class="MuiDrawer-root makeStyles-drawer-94 MuiDrawer-modal"
    data-testid="drawer"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation16 MuiDrawer-paper makeStyles-drawerPaper-95 MuiDrawer-paperAnchorTop"
      style="webkit-transform: none; transform: none; webkit-transition: -webkit-transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        class="MuiGrid-root makeStyles-appButtonContainer-96 MuiGrid-container MuiGrid-spacing-xs-2"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <button
            class="MuiButtonBase-root makeStyles-appButton-288 makeStyles-appButton-290"
            tabindex="0"
            type="button"
          >
            <div
              class="MuiBox-root MuiBox-root-292"
            >
              <div
                class="MuiBox-root MuiBox-root-293 makeStyles-appButtonIcon-289"
              />
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                Marketplace
              </p>
            </div>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <button
            class="MuiButtonBase-root makeStyles-appButton-288 makeStyles-appButton-324"
            tabindex="0"
            type="button"
          >
            <div
              class="MuiBox-root MuiBox-root-325"
            >
              <div
                class="MuiBox-root MuiBox-root-326 makeStyles-appButtonIcon-289"
              />
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                Adjustments
              </p>
            </div>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <button
            class="MuiButtonBase-root makeStyles-appButton-288 makeStyles-appButton-327"
            tabindex="0"
            type="button"
          >
            <div
              class="MuiBox-root MuiBox-root-328"
            >
              <div
                class="MuiBox-root MuiBox-root-329 makeStyles-appButtonIcon-289"
              />
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                Customer
              </p>
            </div>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <button
            class="MuiButtonBase-root makeStyles-appButton-288 makeStyles-appButton-330"
            tabindex="0"
            type="button"
          >
            <div
              class="MuiBox-root MuiBox-root-331"
            >
              <div
                class="MuiBox-root MuiBox-root-332 makeStyles-appButtonIcon-289"
              />
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                Backoffice
              </p>
            </div>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <button
            class="MuiButtonBase-root makeStyles-appButton-288 makeStyles-appButton-333"
            tabindex="0"
            type="button"
          >
            <div
              class="MuiBox-root MuiBox-root-334"
            >
              <div
                class="MuiBox-root MuiBox-root-335 makeStyles-appButtonIcon-289"
              />
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                WebPOS
              </p>
            </div>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="MuiBox-root MuiBox-root-336"
      >
        <header
          class="MuiPaper-root MuiPaper-elevation4 MuiAppBar-root MuiAppBar-positionRelative makeStyles-bar-97 makeStyles-bar-337 MuiAppBar-colorPrimary"
        >
          <div
            class="MuiToolbar-root MuiToolbar-regular makeStyles-toolbar-98 makeStyles-toolbar-338 MuiToolbar-gutters"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root"
              data-testid="mainButton"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  data-testid="Marketplace"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 16 16"
                >
                  <path
                    d="M0 4H4V0H0V4ZM6 16H10V12H6V16ZM4 16H0V12H4V16ZM0 10H4V6H0V10ZM10 10H6V6H10V10ZM12 0V4H16V0H12ZM10 4H6V0H10V4ZM12 10H16V6H12V10ZM16 16H12V12H16V16Z"
                    fill="currentColor"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </header>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
